### ISSUE TITLE
Implement state estimation algorithm 

### ISSUE THREAD

**Issue by ragingonyx** [gt-marine-robotics-group/Virtuoso#1]
As part of the **localization server**, we need a kalman filter to obtain accurate **state estimation** in a noisy environment. This is most easily done with an **Extended Kalman Filter**, though this can in theory be any algorithm for state estimation. Some of the others worth looking into are **Unscented Kalman Filters** and **Factor graphs**. 

The priority is getting any of these algorithms working ASAP so that we can system test other aspects of the architecture that rely on state estimation.

The scope of this issue only covers the development of the algorithm and not the deployment within ROS, which will be a separate project. We want to focus on getting the core logic correct and working before we integrate with the overall system.

Some good stating **documentation** can be found here:

[Paper](https://groups.seas.harvard.edu/courses/cs281/papers/unscented.pdf) outlining differences between EKF and UKF and the math behind it.
[Page](https://towardsdatascience.com/the-unscented-kalman-filter-anything-ekf-can-do-i-can-do-it-better-ce7c773cf88d) explaining in simpler terms what a UKF is and how it's useful compared to a EKF.
[Existing implementation](url) of a basic kalman filter in ROS2.

**Comment by startrek1wars** [comment]
https://www3.nd.edu/~lemmon/courses/ee67033/pubs/julier-ukf-tac-2000.pdf

See this paper for a suggestion on how to keep the covariance matrix positive definite

**Comment by startrek1wars** [comment]
https://github.com/methylDragon/ros-sensor-fusion-tutorial/blob/master/01%20-%20ROS%20and%20Sensor%20Fusion%20Tutorial.md
http://docs.ros.org/en/jade/api/robot_localization/html/index.html
https://automaticaddison.com/set-up-the-odometry-for-a-simulated-mobile-robot-in-ros-2/
Useful robot localization tools.

https://pyproj4.github.io/pyproj/stable/index.html

Might use if implementing custom lat/long to x/y

http://wiki.ros.org/hector_gazebo_plugins
This is the imu/gps plugin vrx uses
